File: python/tradingflow/predictor/variance/shrinkage.py
from enum import IntEnum

import numpy as np

from ._base import VariancePredictor, covariance_predictor
from ._common import (
    correlation_from_covariance,
    sample_covariance,
    schafer_strimmer_alpha,
    single_index_covariance,
)


class Target(IntEnum):
    r"""Shrinkage target selector — the three surveyed in Pantaleo et al.
    (2010), Section III.D.

    - `COMMON_COVARIANCE`: diagonal is the average sample variance,
      off-diagonal the average sample covariance.
    - `CONSTANT_CORRELATION`: diagonal is the sample variances, off-diagonal
      the average off-diagonal correlation times the outer product of the
      standard deviations.
    - `SINGLE_INDEX`: the single-index factor-model covariance.
    """

    COMMON_COVARIANCE = 1
    CONSTANT_CORRELATION = 2
    SINGLE_INDEX = 3


def common_covariance_target(y: np.ndarray, s: np.ndarray) -> np.ndarray:
    n = s.shape[0]
    off = ~np.eye(n, dtype=bool)
    avg_var = float(np.mean(np.diag(s)))
    target = np.full((n, n), float(s[off].mean()) if n > 1 else avg_var)
    np.fill_diagonal(target, avg_var)
    return target


def constant_correlation_target(y: np.ndarray, s: np.ndarray) -> np.ndarray:
    n = s.shape[0]
    corr, stds = correlation_from_covariance(s)
    off = ~np.eye(n, dtype=bool)
    r_bar = float(corr[off].mean()) if n > 1 else 1.0
    target = r_bar * np.outer(stds, stds)
    np.fill_diagonal(target, np.diag(s))
    return target


TARGETS = {
    Target.COMMON_COVARIANCE: common_covariance_target,
    Target.CONSTANT_CORRELATION: constant_correlation_target,
    Target.SINGLE_INDEX: lambda y, s: single_index_covariance(y),
}


def fit(y: np.ndarray, *, target: Target) -> np.ndarray:
    r"""Linear shrinkage \(\Sigma = \alpha F + (1 - \alpha) S\) of the sample
    covariance \(S\) toward a structured target \(F\), with the intensity
    \(\alpha\) estimated analytically by the Schäfer-Strimmer (2005)
    element-wise unbiased estimator."""
    s, centered, finite = sample_covariance(y)
    f = TARGETS[target](y, s)
    alpha, _ = schafer_strimmer_alpha(s, f, centered, finite)
    return alpha * f + (1.0 - alpha) * s


def build(*, target: int | Target = Target.COMMON_COVARIANCE, **kwargs) -> VariancePredictor:
    """Constructs a linear-shrinkage covariance predictor."""
    target = Target(target)
    return covariance_predictor(lambda y: fit(y, target=target), **kwargs)

File: crates/tradingflow/src/operators/predictor/variance/shrinkage.rs
use pyo3::types::PyDictMethods;

use super::VariancePredictor;
use crate::operators::predictor::Config;
use crate::python::py_operator_module;

/// What [`shrinkage`] pulls the sample covariance toward — the three targets
/// surveyed in Pantaleo et al. (2010), Section III.D.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Target {
    /// Every variance the average sample variance, every covariance the
    /// average sample covariance. The most aggressive of the three: it keeps
    /// only the overall level of risk and comovement.
    CommonCovariance = 1,
    /// Sample variances kept, every correlation the average off-diagonal
    /// correlation. Retains the cross-sectional spread of volatility, which
    /// the common-covariance target discards.
    ConstantCorrelation = 2,
    /// The [`single_index`](super::single_index) factor covariance, so the
    /// structure imposed is a market factor rather than a flat constant.
    SingleIndex = 3,
}

/// Linear shrinkage of the sample covariance toward a structured target:
/// `Σ = αF + (1 - α)S`.
///
/// The sample covariance `S` is noisy but unbiased; the target `F` is heavily
/// biased but barely noisy. Mixing them beats either, and the Schäfer-Strimmer
/// (2005) estimator picks the intensity `α` analytically — no cross-validation
/// and no tuning parameter — by comparing the estimated variance of each
/// sample covariance entry against how far it sits from the target. A short
/// window shrinks hard, a long one barely at all.
///
/// See [module-level docs](super::super) for inputs and outputs.
pub fn shrinkage(config: Config, target: Target) -> impl VariancePredictor {
    py_operator_module(
        "tradingflow.predictor.variance.shrinkage",
        config.params(|d| d.set_item("target", target as u8)),
    )
}
